zipline/_protocol.pyx

```diff
@@ -76,6 +79,289 @@ cdef class BarData:
+
+        Parameters
+        ----------
+        assets: Asset or iterable of assets
+
+        Returns
+        -------
+        boolean or Series of booleans, indexed by asset.
+        """
+        dt = self.simulation_dt_func()
+        data_portal = self.data_portal
+
+        if isinstance(assets, Asset):
+            return self._is_stale_for_asset(assets, dt, data_portal)
+        else:
+            return pd.Series(data={
+                asset: self._is_stale_for_asset(asset, dt, data_portal)
+                for asset in assets
+            })
+
+    cdef _is_stale_for_asset(self, asset, dt, data_portal):
+        if asset.start_date > dt:
+            return False
+
+        if asset.end_date <= dt:
+            return False
+
+        current_val = data_portal.get_spot_value(asset, "close", dt,
+                                                 self.data_frequency)
+
+        if not np.isnan(current_val):
+            # found a current value, so we know this asset is not stale.
+            return False
+        else:
+            # if we don't have a current val, we need to distinguish between
+            # if this asset has ever traded (stale = True) or has never traded
+            # (stale = False)
+            last_traded_dt = \
+                data_portal.get_spot_value(asset, "last_traded", dt,
+                                           self.data_frequency)
+
+            return not (last_traded_dt is pd.NaT)
+
+    def history(self, assets, fields, bar_count, frequency):
+        """
+        Returns a window of data for the given assets and fields.
+
+        This data is adjusted for splits, dividends, and mergers as of the
+        current algorithm time.
+
+        The semantics of missing data are identical to the ones described in
+        the notes for `get_spot_value`.
+
+        Parameters
+        ----------
+        assets: Asset or iterable of Asset
+
+        fields: string or iterable of string.  Valid values are "open", "high",
+            "low", "close", "volume", "price", and "last_traded".
+
+        bar_count: integer number of bars of trade data
+
+        frequency: string. "1m" for minutely data or "1d" for daily date
+
+        Returns
+        -------
+        series or DataFrame or Panel, depending on the dimensionality of
+            the 'assets' and 'fields' parameters.
+
+            If single asset and field are passed in, the returned Series is
+            indexed by dt.
+
+            If multiple assets and single field are passed in, the returned
+            DataFrame is indexed by dt, and has assets as columns.
+
+            If a single asset and multiple fields are passed in, the returned
+            DataFrame is indexed by dt, and has fields as columns.
+
+            If multiple assets and multiple fields are passed in, the returned
+            Panel is indexed by field, has dt as the major axis, and assets
+            as the minor axis.
```

Comment:
Does this table summarize this correctly?


|                  | scalar Asset                         | iterable Assets                                         |
|----------------|---------------------------------------|--------------------------------------------------------|
| scalar field | scalar (float or int)                   | `DataFrame(index=dts, columns=assets)                      |
| iterable fields | `DataFrame(index=dts, columns=fields)` | `Panel(index=fields, major_axis=dt, minor_axis=assets) `    |